### ISSUE TITLE
Figure out test infrastructure

### ISSUE THREAD

**Issue by Gozala** [mozilla/libdweb#12]
Unfortunately as far as I understand `web-ext` tool does not come with test infrastructure for us to use to exercise developed WebExtension APIs. Creating this bug to track the research on how can we do this without reinventing everything.

**Comment by Gozala** [comment]
@rpl You mentioned you've done something along this lines with marionette. Could you please provide some pointers ?

**Comment by rpl** [comment]
@Gozala here is some links related to the test strategy that I was showing you when we met at the All Hands:
- https://github.com/mozilla/webextension-polyfill/tree/master/test/integration
- https://github.com/mozilla/webextension-polyfill/tree/master/test/fixtures

This is the strategy that I recently chose for the webextension-polyfill, but the main goal of that test suite is to test the behavior of the library across the supported browsers (basically Chrome as the main target, and Firefox for comparison with the native behaviors), and so it is currently limited to what can be tested without access to the privileged part of the browser UI (e.g. there is no way to trigger a browserAction popup or to answer to a optional permission prompt etc.)

Nevertheless, if the tests only need to run on Firefox (like in this case), this is a limitation that can be overcome easily by writing some helpers that run privileged js code using driver.setContext(Context.CHROME); (`Context.CHROME` is provided by "selenium-webdriver/firefox").

@standard8 put together a very nice example of this approach in this repository:

- https://github.com/Standard8/example-webextension/blob/4396e009f8e3221a5d5b38d2fc04fdc4a65cf871/test/functional/utils.js#L59-L104

**Comment by Gozala** [comment]
## Status update

- [x] I have set few things up in #25 based on suggestions above.
- [x] Now I can lunch nightly with tests pre-installed and I'm able to exercise an API
- [ ] I'd like to make [blue-tape](https://www.npmjs.com/package/blue-tape) available with-in the extension context to define unit tests.
- [ ] I'd like to observe blue-tape output in the node process for reporting, but so far have no luck getting any output through webdriver. I could probably borrow code from [jpm test](https://github.com/mozilla-jetpack/) but ideally I would not diverge from WebExtensions ecosystem too much.

**Comment by Gozala** [comment]
It seems that [Extension pages](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/user_interface/Extension_pages) could be most reasonable way to go about testing. We could probably bundle up tests the same way as tape tests are run in a content page and then use puppeteer or a webdriver to open tab for add-on page.

Alternative would be to add `test` API and use [fx-runner](https://github.com/mozilla-jetpack/node-fx-runner) to do the rest.

**Comment by Gozala** [comment]
I have landed some changes that add's `test` API to the libdweb that implements `tape` interface. `selenium-webdriver` is used to start firefox nightly and install temporary test web-extension (as per @rpl examples). `stdout` from firefox is then consumed to emit TAP output and to figure out when tests are complete, after which process exits with either `0` if all tests passed or `1` if any test failed.

**Complication**: Sadly there `selenium-driver` does not provide a way to consume browser `stdout`,  instead they provide a way to configure `stdio` for it via [`DriverService.Builder.prototype.setStdio`](https://seleniumhq.github.io/selenium/docs/api/javascript/module/selenium-webdriver/remote/index_exports_DriverService.Builder.html#setStdio) but since [spawn process in never exposed](https://github.com/SeleniumHQ/selenium/blob/5f52b02163ce12ec6612d58a6989f45f064ac7ea/javascript/node/selenium-webdriver/io/exec.js#L128-L161) the only way I managed to get browser output is by [forwarding it to a temp file](https://github.com/mozilla/libdweb/blob/97d2229b7278a76c7828d1124af2c58741511383/test/setup.js#L53-L69) and then (watching / reading from it)[https://github.com/mozilla/libdweb/blob/97d2229b7278a76c7828d1124af2c58741511383/test/setup.js#L89-L101) which sadly turned out very unreliable (often even though output is written I don't get notified).

I have submitted SeleniumHQ/selenium#6204 with a request to provide access to `stdout` / `stderror` but in the meantime I'm happy to get some suggestions / help in trying to get this setup more reliable.

**Comment by Gozala** [comment]
End up avoiding issue described in the above comment by using node cluster api which runs selenium in the worker process while stdout is analyzed from the master.

This also sets a foundation for the future change were I'd like to use master for traversing test directory and dispatching test to the worker.

**Comment by Gozala** [comment]
I have took things I've put together for running tests and created https://github.com/gozala/tape-ext which then is used here to execute test extensions. I think it works reasonably well enough and provides reasonable flow for authoring tests so I'll be closing this.

One more thing I still wish we were able to do is author tests where server is in node and client is in firefox and vice versa. Although technically possible it would require significant effort and I'm not sure it's worth it, especially since it would complicate things if we were to land this into firefox tree.
